### PR TITLE
Fixing NPE in getCurrentAuditor

### LINKED ISSUE #4861: NullPointer throw when authenticate via Bearer Token to active endpoint i.e. register application
I've configured SCDF to use ADFS nad OAuth2.0. When user login to SCDF everything is fine, but when I want to call REST API to register new application I've got NullPointer:

`curl -vik -H "Authorization: Bearer %TOKEN%" https://localhost:9445/apps/task/my-task-process/1.7.4.90-SNAPSHOT -d "uri=docker%3Aeu.gcr.io%2Fmy-project-gcp%2Fmy-task-process%3A1.7.4.90-SNAPSHOT&force=true"`

While debug I noticed that it's caused by getCurrentAuditor method because when I authorize via client, there is no user info. 

```
2022-03-23 14:48:04.419 TRACE 5132 --- [nio-9445-exec-9] o.s.s.w.a.i.FilterSecurityInterceptor    : Did not re-authenticate JwtAuthenticationToken [Principal=org.springframework.security.oauth2.jwt.Jwt@6bc00fb2, Credentials=[PROTECTED], Authenticated=true, Details=WebAuthenticationDetails [RemoteIpAddress=127.0.0.1, SessionId=null], Granted Authorities=[ROLE_CREATE, ROLE_MANAGE, ROLE_DEPLOY, ROLE_MODIFY, ROLE_DESTROY, ROLE_SCHEDULE, ROLE_VIEW]] before authorizing
2022-03-23 14:48:04.419 TRACE 5132 --- [nio-9445-exec-9] o.s.s.w.a.i.FilterSecurityInterceptor    : Authorizing filter invocation [POST /apps/task/my-task-process/1.7.4.90-SNAPSHOT] with attributes [hasRole('ROLE_CREATE')]
2022-03-23 14:48:04.419 DEBUG 5132 --- [nio-9445-exec-9] o.s.s.w.a.i.FilterSecurityInterceptor    : Authorized filter invocation [POST /apps/task/my-task-process/1.7.4.90-SNAPSHOT] with attributes [hasRole('ROLE_CREATE')]
2022-03-23 14:48:04.419 TRACE 5132 --- [nio-9445-exec-9] o.s.s.w.a.i.FilterSecurityInterceptor    : Did not switch RunAs authentication since RunAsManager returned null
2022-03-23 14:48:04.419 DEBUG 5132 --- [nio-9445-exec-9] o.s.security.web.FilterChainProxy        : Secured POST /apps/task/my-task-process/1.7.4.90-SNAPSHOT
2022-03-23 14:48:46.327 ERROR 5132 --- [nio-9445-exec-9] o.s.c.d.s.c.RestControllerAdvice         : Caught exception while handling a request

java.lang.NullPointerException: null
        at java.base/java.util.Objects.requireNonNull(Objects.java:221)
        at java.base/java.util.Optional.<init>(Optional.java:107)
        at java.base/java.util.Optional.of(Optional.java:120)
        at org.springframework.cloud.dataflow.server.service.SpringSecurityAuditorAware.getCurrentAuditor(SpringSecurityAuditorAware.java:45)
        at java.base/java.util.Optional.map(Optional.java:265)
        at org.springframework.data.auditing.AuditingHandler.getAuditor(AuditingHandler.java:109)
        at org.springframework.data.auditing.AuditingHandler.markCreated(AuditingHandler.java:92)
        at org.springframework.data.jpa.domain.support.AuditingEntityListener.touchForCreate(AuditingEntityListener.java:92)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.hibernate.jpa.event.internal.ListenerCallback.performCallback(ListenerCallback.java:35)
        at org.hibernate.jpa.event.internal.CallbackRegistryImpl.callback(CallbackRegistryImpl.java:95)
        at org.hibernate.jpa.event.internal.CallbackRegistryImpl.preCreate(CallbackRegistryImpl.java:55)
        at org.hibernate.event.internal.AbstractSaveEventListener.saveWithGeneratedId(AbstractSaveEventListener.java:108)
```

### PR DESCRIPTION
**Pull request by corneil**
Fixing NPE in getCurrentAuditor.

Fixes #4861

### CODE REVIEW

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/SpringSecurityAuditorAware.java [hunk 1]
```diff
@@ -41,8 +43,8 @@ public SpringSecurityAuditorAware(SecurityStateBean securityStateBean) {
 		final boolean authenticationEnabled = securityStateBean.isAuthenticationEnabled();
 		if (authenticationEnabled && SecurityContextHolder.getContext() != null) {
 			final Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
-			if (!(authentication instanceof AnonymousAuthenticationToken)) {
-				return Optional.of(authentication.getName());
+			if (authentication != null && !(authentication instanceof AnonymousAuthenticationToken)) {
```
Comment (onobc) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/SpringSecurityAuditorAware.java:46]: I believe this null check is redundant as `null instanceof Foo)` is always `false`.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/SpringSecurityAuditorAware.java [hunk 2]
```diff
@@ -41,8 +43,8 @@ public SpringSecurityAuditorAware(SecurityStateBean securityStateBean) {
...
+				return Optional.ofNullable(authentication.getName());
```
Comment (onobc) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/SpringSecurityAuditorAware.java:47]: Do you mind adding a test to verify this case?

**Review by onobc — CHANGES_REQUESTED** [review]
https://github.com/spring-cloud/spring-cloud-dataflow/pull/4876/files#r844296872

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/SpringSecurityAuditorAware.java (hunk 1 above)
Comment (jvalkeal) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/SpringSecurityAuditorAware.java:46]: Ah there's the `!`, which then needs a null check.

Comment (onobc) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/SpringSecurityAuditorAware.java:46]: Touche' ..... LOL great catch

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/service/SpringSecurityAuditorAwareTests.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+import org.springframework.security.core.context.SecurityContext;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.oauth2.core.OAuth2AccessToken;
+import org.springframework.security.oauth2.server.resource.authentication.BearerTokenAuthentication;
+import org.springframework.security.oauth2.server.resource.introspection.OAuth2IntrospectionAuthenticatedPrincipal;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Tests for org.springframework.cloud.dataflow.server.service.SpringSecurityAuditorAware
```
Comment (onobc) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/service/SpringSecurityAuditorAwareTests.java:37]: **[NIT]** Using `@{link` makes the doc a bit cleaner such as: 
```java
/** 
 * Unit tests for {@link SpringSecurityAuditorAware}.
```

I can adjust during merge though.

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/service/SpringSecurityAuditorAwareTests.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+ * @author Corneil du Plessis
+ */
+public class SpringSecurityAuditorAwareTests {
+	/**
+	 * Test the fix for the case where the authentication doesn't contain a name but isn't anonymous it resulted in an NPE.
+	 */
+	@Test
+	public void testNullPointerOnUnnamedPrincipal() {
```
Comment (onobc) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/service/SpringSecurityAuditorAwareTests.java:45]: **[NIT]** 2 features Junit5 brings that we should use more is 
1) no need for public class or methods
2) no need to prefix method names w/ "test"

I will adjust during merge.

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/service/SpringSecurityAuditorAwareTests.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+		// given
+		HashMap<String, Object> attributes = new HashMap<>();
+		attributes.put("attr1", "value1");
+		final OAuth2IntrospectionAuthenticatedPrincipal principal = new OAuth2IntrospectionAuthenticatedPrincipal(
```
Comment (onobc) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/service/SpringSecurityAuditorAwareTests.java:49]: Typically we don't use final unless there is a good reason.

**Comment by onobc** [comment]
Closed via https://github.com/spring-cloud/spring-cloud-dataflow/commit/b8bf2ded9ffd3b4b402329db2fda96bece413f8d

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/service/SpringSecurityAuditorAwareTests.java (hunk 3 above)
Comment (corneil) [reply spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/service/SpringSecurityAuditorAwareTests.java:37]: Thanks for the tip.

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/service/SpringSecurityAuditorAwareTests.java (hunk 5 above)
Comment (corneil) [reply spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/service/SpringSecurityAuditorAwareTests.java:49]: Only the authentication need to be final

**Comment by corneil** [comment]
And !false is true.

**Pull request closed without merging**